src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java

```diff
@@ -159,6 +168,33 @@ void rollback(DeploymentDocument deploymentDocument, CompletableFuture<Deploymen
         }
     }
 
+    /**
+     * Finds all services which are dependers of given broken services, directly or indirectly
+     * This method performs a breadth-first search, starting from the broken services and traversing through
+     * service dependencies.
+     * @param rollbackServices the set of rollback services to track
+     * @param brokenServiceNames the set of broken service names
+     * @return a set of all services depending on the broken services, including themselves
+     */
+    private Set<GreengrassService> findServiceDependers(final Set<GreengrassService> rollbackServices,
+                                                        final Set<String> brokenServiceNames) {
+        LinkedHashSet<GreengrassService> dependerServices = rollbackServices.stream()
+                .filter(service -> brokenServiceNames.contains(service.getName()))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
```

Comment:
why do we need a linked hashset? Just do `toSet`?

Reply:
Because I want to make sure the traverse is in an ordered way, but I think using set is also fine since we only need to make sure all dependers are processed and we're using set for de-duplication . Just curious why we always prefer to use basic data-structure? (saving space?)

Reply:
Does the order matter? By doing something other than `toSet` you're implying that the order is meaningful (or you're using a specific Set type for some other reason).

Reply:
Updated